pinot-query-planner/src/main/java/org/apache/pinot/query/planner/explain/PlanNodeSorter.java

```diff
@@ -0,0 +1,228 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.query.planner.explain;
+
+import com.google.common.collect.Sets;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.TreeSet;
+import org.apache.pinot.common.proto.Plan;
+import org.apache.pinot.core.query.reduce.ExplainPlanDataTableReducer;
+import org.apache.pinot.query.planner.plannode.AggregateNode;
+import org.apache.pinot.query.planner.plannode.ExchangeNode;
+import org.apache.pinot.query.planner.plannode.ExplainedNode;
+import org.apache.pinot.query.planner.plannode.FilterNode;
+import org.apache.pinot.query.planner.plannode.JoinNode;
+import org.apache.pinot.query.planner.plannode.MailboxReceiveNode;
+import org.apache.pinot.query.planner.plannode.MailboxSendNode;
+import org.apache.pinot.query.planner.plannode.PlanNode;
+import org.apache.pinot.query.planner.plannode.PlanNodeVisitor;
+import org.apache.pinot.query.planner.plannode.ProjectNode;
+import org.apache.pinot.query.planner.plannode.SetOpNode;
+import org.apache.pinot.query.planner.plannode.SortNode;
+import org.apache.pinot.query.planner.plannode.TableScanNode;
+import org.apache.pinot.query.planner.plannode.ValueNode;
+import org.apache.pinot.query.planner.plannode.WindowNode;
+
+
+/**
+ * A utility class used to sort the plan nodes in a deterministic order.
+ *
+ * Any comparator can be passed to the sort method to sort the plan nodes, although the default comparator
+ * is used to sort the plan nodes based on the type and the attributes of the node.
+ */
+public class PlanNodeSorter {
```

Comment:
I didn't quite get the purpose of this plan node sorter utility. In which cases would we get a non-deterministic order of plan nodes (necessitating this plan node sorter)?

Reply:
We receive different plans from different servers and each server generate a different plan for each segment. The order inside each server is not guaranteed (they are calculated concurrently). Once we receive the server plans in the broker, we first sort each server plan.

Once we have each server plan sorted, we merge them easily. Not only that, the result will always be shown in the same, which will help users. Otherwise two explains of the same query on the same data will return plans that are semantically equal but the order of some nodes will be different.

Remember this sorting is only applied to `Combine` explain nodes. In other cases reordering inputs may break semantics.

Two notes:
1. This is something we also do in single-stage explain (see ExplainPlanDataTableReducer). In fact I copied the definition from there. As you can imagine, just sorting the nodes whose name contains `Combine` is not formally correct, but seems to be good enough to be used for a while.
2. Your comment make me realize the code right now is looking for `COMBINE` in the text instead of the new upper camel case `Combine`. I'm changing that in a new commit.